blib/man3/Tk.X.3
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
'''   \*(M", \*(S", \*(N" and \*(T" are the equivalent of
'''   \*(L" and \*(R", except that they are used on ".xx" lines,
'''   such as .IP and .SH, which do another additional levels of
'''   double-quote interpretation
.ds M" """
.ds S" """
.ds N" """""
.ds T" """""
.ds L' '
.ds R' '
.ds M' '
.ds S' '
.ds N' '
.ds T' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds M" ``
.ds S" ''
.ds N" ``
.ds T" ''
.ds L' `
.ds R' '
.ds M' `
.ds S' '
.ds N' `
.ds T' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH X 3 "Tk402.001" "26/Mar/97" "perl/Tk Documentation"
.IX Title "X 3"
.UC
.IX Name "Tk::X - Perl extension for Xlib constants."
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
Tk::X \- Perl extension for Xlib constants.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.PP
.Vb 2
\&  use Tk::X;
\&  
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
A module generated by h2xs. It exists to export Xlib #define type
constants for possible use with Tk::Xlib.
.SH "Exported constants"
.IX Header "Exported constants"
.PP
.Vb 346
\&  Above
\&  AllTemporary
\&  AllocAll
\&  AllocNone
\&  AllowExposures
\&  AlreadyGrabbed
\&  Always
\&  AnyButton
\&  AnyKey
\&  AnyModifier
\&  AnyPropertyType
\&  ArcChord
\&  ArcPieSlice
\&  AsyncBoth
\&  AsyncKeyboard
\&  AsyncPointer
\&  AutoRepeatModeDefault
\&  AutoRepeatModeOff
\&  AutoRepeatModeOn
\&  BadAccess
\&  BadAlloc
\&  BadAtom
\&  BadColor
\&  BadCursor
\&  BadDrawable
\&  BadFont
\&  BadGC
\&  BadIDChoice
\&  BadImplementation
\&  BadLength
\&  BadMatch
\&  BadName
\&  BadPixmap
\&  BadRequest
\&  BadValue
\&  BadWindow
\&  Below
\&  BottomIf
\&  Button1
\&  Button1Mask
\&  Button1MotionMask
\&  Button2
\&  Button2Mask
\&  Button2MotionMask
\&  Button3
\&  Button3Mask
\&  Button3MotionMask
\&  Button4
\&  Button4Mask
\&  Button4MotionMask
\&  Button5
\&  Button5Mask
\&  Button5MotionMask
\&  ButtonMotionMask
\&  ButtonPress
\&  ButtonPressMask
\&  ButtonRelease
\&  ButtonReleaseMask
\&  CWBackPixel
\&  CWBackPixmap
\&  CWBackingPixel
\&  CWBackingPlanes
\&  CWBackingStore
\&  CWBitGravity
\&  CWBorderPixel
\&  CWBorderPixmap
\&  CWBorderWidth
\&  CWColormap
\&  CWCursor
\&  CWDontPropagate
\&  CWEventMask
\&  CWHeight
\&  CWOverrideRedirect
\&  CWSaveUnder
\&  CWSibling
\&  CWStackMode
\&  CWWidth
\&  CWWinGravity
\&  CWX
\&  CWY
\&  CapButt
\&  CapNotLast
\&  CapProjecting
\&  CapRound
\&  CenterGravity
\&  CirculateNotify
\&  CirculateRequest
\&  ClientMessage
\&  ClipByChildren
\&  ColormapChangeMask
\&  ColormapInstalled
\&  ColormapNotify
\&  ColormapUninstalled
\&  Complex
\&  ConfigureNotify
\&  ConfigureRequest
\&  ControlMapIndex
\&  ControlMask
\&  Convex
\&  CoordModeOrigin
\&  CoordModePrevious
\&  CopyFromParent
\&  CreateNotify
\&  CurrentTime
\&  CursorShape
\&  DefaultBlanking
\&  DefaultExposures
\&  DestroyAll
\&  DestroyNotify
\&  DirectColor
\&  DisableAccess
\&  DisableScreenInterval
\&  DisableScreenSaver
\&  DoBlue
\&  DoGreen
\&  DoRed
\&  DontAllowExposures
\&  DontPreferBlanking
\&  EastGravity
\&  EnableAccess
\&  EnterNotify
\&  EnterWindowMask
\&  EvenOddRule
\&  Expose
\&  ExposureMask
\&  FamilyChaos
\&  FamilyDECnet
\&  FamilyInternet
\&  FillOpaqueStippled
\&  FillSolid
\&  FillStippled
\&  FillTiled
\&  FirstExtensionError
\&  FocusChangeMask
\&  FocusIn
\&  FocusOut
\&  FontChange
\&  FontLeftToRight
\&  FontRightToLeft
\&  ForgetGravity
\&  GCArcMode
\&  GCBackground
\&  GCCapStyle
\&  GCClipMask
\&  GCClipXOrigin
\&  GCClipYOrigin
\&  GCDashList
\&  GCDashOffset
\&  GCFillRule
\&  GCFillStyle
\&  GCFont
\&  GCForeground
\&  GCFunction
\&  GCGraphicsExposures
\&  GCJoinStyle
\&  GCLastBit
\&  GCLineStyle
\&  GCLineWidth
\&  GCPlaneMask
\&  GCStipple
\&  GCSubwindowMode
\&  GCTile
\&  GCTileStipXOrigin
\&  GCTileStipYOrigin
\&  GXand
\&  GXandInverted
\&  GXandReverse
\&  GXclear
\&  GXcopy
\&  GXcopyInverted
\&  GXequiv
\&  GXinvert
\&  GXnand
\&  GXnoop
\&  GXnor
\&  GXor
\&  GXorInverted
\&  GXorReverse
\&  GXset
\&  GXxor
\&  GrabFrozen
\&  GrabInvalidTime
\&  GrabModeAsync
\&  GrabModeSync
\&  GrabNotViewable
\&  GrabSuccess
\&  GraphicsExpose
\&  GravityNotify
\&  GrayScale
\&  HostDelete
\&  HostInsert
\&  IncludeInferiors
\&  InputFocus
\&  InputOnly
\&  InputOutput
\&  IsUnmapped
\&  IsUnviewable
\&  IsViewable
\&  JoinBevel
\&  JoinMiter
\&  JoinRound
\&  KBAutoRepeatMode
\&  KBBellDuration
\&  KBBellPercent
\&  KBBellPitch
\&  KBKey
\&  KBKeyClickPercent
\&  KBLed
\&  KBLedMode
\&  KeyPress
\&  KeyPressMask
\&  KeyRelease
\&  KeyReleaseMask
\&  KeymapNotify
\&  KeymapStateMask
\&  LASTEvent
\&  LSBFirst
\&  LastExtensionError
\&  LeaveNotify
\&  LeaveWindowMask
\&  LedModeOff
\&  LedModeOn
\&  LineDoubleDash
\&  LineOnOffDash
\&  LineSolid
\&  LockMapIndex
\&  LockMask
\&  LowerHighest
\&  MSBFirst
\&  MapNotify
\&  MapRequest
\&  MappingBusy
\&  MappingFailed
\&  MappingKeyboard
\&  MappingModifier
\&  MappingNotify
\&  MappingPointer
\&  MappingSuccess
\&  Mod1MapIndex
\&  Mod1Mask
\&  Mod2MapIndex
\&  Mod2Mask
\&  Mod3MapIndex
\&  Mod3Mask
\&  Mod4MapIndex
\&  Mod4Mask
\&  Mod5MapIndex
\&  Mod5Mask
\&  MotionNotify
\&  NoEventMask
\&  NoExpose
\&  NoSymbol
\&  Nonconvex
\&  None
\&  NorthEastGravity
\&  NorthGravity
\&  NorthWestGravity
\&  NotUseful
\&  NotifyAncestor
\&  NotifyDetailNone
\&  NotifyGrab
\&  NotifyHint
\&  NotifyInferior
\&  NotifyNonlinear
\&  NotifyNonlinearVirtual
\&  NotifyNormal
\&  NotifyPointer
\&  NotifyPointerRoot
\&  NotifyUngrab
\&  NotifyVirtual
\&  NotifyWhileGrabbed
\&  Opposite
\&  OwnerGrabButtonMask
\&  ParentRelative
\&  PlaceOnBottom
\&  PlaceOnTop
\&  PointerMotionHintMask
\&  PointerMotionMask
\&  PointerRoot
\&  PointerWindow
\&  PreferBlanking
\&  PropModeAppend
\&  PropModePrepend
\&  PropModeReplace
\&  PropertyChangeMask
\&  PropertyDelete
\&  PropertyNewValue
\&  PropertyNotify
\&  PseudoColor
\&  RaiseLowest
\&  ReparentNotify
\&  ReplayKeyboard
\&  ReplayPointer
\&  ResizeRedirectMask
\&  ResizeRequest
\&  RetainPermanent
\&  RetainTemporary
\&  RevertToNone
\&  RevertToParent
\&  RevertToPointerRoot
\&  ScreenSaverActive
\&  ScreenSaverReset
\&  SelectionClear
\&  SelectionNotify
\&  SelectionRequest
\&  SetModeDelete
\&  SetModeInsert
\&  ShiftMapIndex
\&  ShiftMask
\&  SouthEastGravity
\&  SouthGravity
\&  SouthWestGravity
\&  StaticColor
\&  StaticGravity
\&  StaticGray
\&  StippleShape
\&  StructureNotifyMask
\&  SubstructureNotifyMask
\&  SubstructureRedirectMask
\&  Success
\&  SyncBoth
\&  SyncKeyboard
\&  SyncPointer
\&  TileShape
\&  TopIf
\&  TrueColor
\&  UnmapGravity
\&  UnmapNotify
\&  Unsorted
\&  VisibilityChangeMask
\&  VisibilityFullyObscured
\&  VisibilityNotify
\&  VisibilityPartiallyObscured
\&  VisibilityUnobscured
\&  WestGravity
\&  WhenMapped
\&  WindingRule
\&  XYBitmap
\&  XYPixmap
\&  X_H
\&  X_PROTOCOL
\&  X_PROTOCOL_REVISION
\&  YSorted
\&  YXBanded
\&  YXSorted
\&  ZPixmap
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
Nick Ing-Simmons ran h2xs ...
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\fITk::Xlib\fR\|(3).

.rn }` ''


blib/lib/Tk.pm
#
# Copyright (c) 1992-1994 The Regents of the University of California.
# Copyright (c) 1994 Sun Microsystems, Inc.
# Copyright (c) 1995-1997 Nick Ing-Simmons. All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself, subject 
# to additional disclaimer in Tk/license.terms due to partial
# derivation from Tk4.0 sources.
#
package Tk;
require 5.003_97;
use     AutoLoader qw(AUTOLOAD);
use     DynaLoader;
require Exporter;
@Tk::ISA = qw(Exporter DynaLoader);


@EXPORT    = qw(Exists Ev after exit MainLoop DoOneEvent tkinit);
@EXPORT_OK = qw(NoOp *widget *event lsearch catch 
                DONT_WAIT WINDOW_EVENTS  FILE_EVENTS TIMER_EVENTS 
                IDLE_EVENTS ALL_EVENTS 
                NORMAL_BG ACTIVE_BG SELECT_BG 
                SELECT_FG TROUGH INDICATOR DISABLED BLACK WHITE);
%EXPORT_TAGS = (eventtypes => [qw(DONT_WAIT WINDOW_EVENTS  FILE_EVENTS 
                                  TIMER_EVENTS IDLE_EVENTS ALL_EVENTS)], 
                variables  => [qw(*widget *event)],
                colors     => [qw(NORMAL_BG ACTIVE_BG SELECT_BG SELECT_FG 
                                  TROUGH INDICATOR DISABLED BLACK WHITE)],
               );

use strict;
use Symbol ();

use Carp;

# $tk_version and $tk_patchLevel are reset by pTk when a mainwindow
# is created, $VERSION is checked by bootstrap
$Tk::version     = "4.2";
$Tk::patchLevel  = "4.2";
$Tk::VERSION     = '402.001';
$Tk::strictMotif = 0;
                                   
{($Tk::library) = __FILE__ =~ /^(.*)\.pm$/;}
$Tk::library = Tk->findINC('.') unless (-d $Tk::library);

$Tk::widget  = undef;
$Tk::event   = undef;

# Supress used once warnings on function table pointers 
# How can we do this in the C code?
use vars qw($TkVtab $TkintVtab $LangVtab $TkglueVtab $XlibVtab $TkoptionVtab);  
use vars qw($TixVtab $TixintVtab $TiximgxpmVtab);

bootstrap Tk $Tk::VERSION;

{
 no strict 'refs';
 *{'exit'} = \&Exit;
}
my $boot_time = timeofday();

# This is a workround for Solaris X11 locale handling 
Preload(DynaLoader::dl_findfile('-L/usr/openwin/lib','-lX11')) if (&NeedPreload && -d '/usr/openwin/lib');

use Tk::Submethods ('option'    =>  [qw(add get clear readfile)],
                    'clipboard' =>  [qw(clear append)]
                   );

sub BackTrace
{
 my $w = shift;
 return unless (@_ || $@);
 my $mess = (@_) ? shift : "$@";
 my $i = 0;  
 my ($pack,$file,$line,$sub) = caller($i++);
 while (1)   
  {          
   my $loc = "at $file line $line";
   ($pack,$file,$line,$sub) = caller($i++);
   last if (!defined($sub) || $sub eq '(eval)');
   $w->AddErrorInfo("$sub $loc");
  }          
 die "$mess\n";
}

sub NoOp  { }

sub Ev
{
 my @args = @_;
 my $obj;
 if (@args == 1)
  {
   my $arg = pop(@args);
   $obj = (ref $arg) ? $arg : \$arg;
  }
 else 
  {
   $obj = \@args;
  }
 return bless $obj,"Tk::Ev";
}

sub InitClass
{
 my ($package,$parent) = @_;
 croak "Unexpected type of parent $parent" unless(ref $parent);
 croak "$parent is not a widget" unless($parent->IsWidget);
 my $mw = $parent->MainWindow;
 unless (exists $mw->{'_ClassInit_'}{$package})
  {
   $package->Install($mw);
   $mw->{'_ClassInit_'}{$package} = $package->ClassInit($mw);
  }
}

require Tk::Widget;
require Tk::Image;
require Tk::MainWindow;

sub Exists
{my $w = shift;
 return defined($w) && ref($w) && $w->IsWidget && $w->exists;
}

sub Time_So_Far
{
 return timeofday() - $boot_time;
} 

# Selection* are not autoloaded as names are too long.

sub SelectionOwn
{my $widget = shift;
 selection('own',(@_,$widget));
}

sub SelectionOwner
{
 selection('own',"-displayof",@_);
}

sub SelectionClear
{
 selection('clear',"-displayof",@_);
}

sub SelectionExists
{
 selection('exists',"-displayof",@_);
}

sub SelectionHandle
{my $widget = shift;
 my $command = pop;
 selection('handle',@_,$widget,$command);
}

#
# This is a $SIG{__DIE__} handler which does not change the $@
# string in the way 'croak' does, but rather add to Tk's ErrorInfo.
# It stops at 1st enclosing eval on assumption that the eval
# is part of Tk call process and will add its own context to ErrorInfo
# and then pass on the error.
# 
sub __DIE__
{
 my $mess = shift;
 my $w = $Tk::widget;
 if (defined $w)
  {
   my $i = 0;  
   my ($pack,$file,$line,$sub) = caller($i++);
   while (1)   
    {          
     my $loc = "at $file line $line";
     ($pack,$file,$line,$sub) = caller($i++);
     last if (!defined($sub) || $sub eq '(eval)');
     $w->AddErrorInfo("$sub $loc");
    }          
  }
}

sub fileevent
{
 require Tk::IO;
 my ($obj,$file,$mode,$cb) = @_;
 croak "Unknown mode '$mode'" unless $mode =~ /^(readable|writeable)$/;
 unless (ref $file)
  {
   require IO::Handle;
   no strict 'refs';
   $file = Symbol::qualify($file,(caller)[0]);
   $file = bless \*{$file},'IO::Handle';
  }
 if ($cb)
  {
   # Adding the handler
   $cb = Tk::Callback->new($cb);
   if ($mode eq 'readable')
    {
     Tk::IO::CreateReadHandler($file,$cb);
    }
   else
    {
     Tk::IO::CreateWriteHandler($file,$cb);
    }
  }
 else
  {
   if ($mode eq 'readable')
    {
     Tk::IO::DeleteReadHandler($file);
    }
   else
    {
     Tk::IO::DeleteWriteHandler($file);
    }
  }
}

sub SplitString
{
 local $_ = shift;
 carp "SplitString '$_'";
 my (@arr, $tmp);
 while (/\{([^{}]*)\}|((?:[^\s\\]|\\.)+)/gs) {
   if (defined $1) { push @arr, $1 }
   else { $tmp = $2 ; $tmp =~ s/\\([\s\\])/$1/g; push @arr, $tmp }
 }
 return @arr;
 #return split(/\s+/,$_);
}


1;

__END__
# provide an exit() to be exported if exit occurs 
# before a MainWindow->new()
sub exit { CORE::exit(@_);}

sub Exists
{my $w = shift;
 return defined($w) && ref($w) && $w->IsWidget && $w->exists;
}

sub Error
{my $w = shift;
 my $error = shift;
 if (Exists($w))
  {
   my $grab = $w->grab('current');  
   $grab->Unbusy if (defined $grab);
  }
 chomp($error);
 warn "Tk::Error: $error\n " . join("\n ",@_);
}

sub tkinit
{
 return MainWindow->new(@_);
}

sub CancelRepeat
{
 my $w = shift->MainWindow;
 my $id = delete $w->{_afterId_};
 $w->after('cancel',$id) if (defined $id);
}

sub RepeatId
{
 my ($w,$id) = @_;
 $w = $w->MainWindow;
 $w->CancelRepeat;
 $w->{_afterId_} = $id;
}



#----------------------------------------------------------------------------
# focus.tcl --
#
# This file defines several procedures for managing the input
# focus.
#
# @(#) focus.tcl 1.6 94/12/19 17:06:46
#
# Copyright (c) 1994 Sun Microsystems, Inc.
#
# See the file "license.terms" for information on usage and redistribution
# of this file, and for a DISCLAIMER OF ALL WARRANTIES.

sub FocusChildren { shift->children }

#
# focusNext --
# This procedure is invoked to move the input focus to the next window
# after a given one. "Next" is defined in terms of the window
# stacking order, with all the windows underneath a given top-level
# (no matter how deeply nested in the hierarchy) considered except
# for frames and toplevels.
#
# Arguments:
# w - Name of a window: the procedure will set the focus
# to the next window after this one in the traversal
# order.
sub focusNext
{
 my $w = shift;
 my $cur = $w;
 while (1)
  {
   # Descend to just before the first child of the current widget.
   my $parent = $cur;
   my @children = $cur->FocusChildren();
   my $i = -1;
   # Look for the next sibling that isn't a top-level.
   while (1)
    {
     $i += 1;
     if ($i < @children)
      {
       $cur = $children[$i];
       next if ($cur->toplevel == $cur);
       last
      }
     # No more siblings, so go to the current widget's parent.
     # If it's a top-level, break out of the loop, otherwise
     # look for its next sibling.
     $cur = $parent;
     last if ($cur->toplevel() == $cur);
     $parent = $parent->parent();
     @children = $parent->FocusChildren();
     $i = lsearch(\@children,$cur);
    }
   if ($cur == $w || $cur->FocusOK)
    {
     $cur->Tk::focus;
     return;
    }
  }
}
# focusPrev --
# This procedure is invoked to move the input focus to the previous
# window before a given one. "Previous" is defined in terms of the
# window stacking order, with all the windows underneath a given
# top-level (no matter how deeply nested in the hierarchy) considered.
#
# Arguments:
# w - Name of a window: the procedure will set the focus
# to the previous window before this one in the traversal
# order.
sub focusPrev
{
 my $w = shift;
 my $cur = $w;
 my @children;
 my $i;
 my $parent;
 while (1)
  {
   # Collect information about the current window's position
   # among its siblings. Also, if the window is a top-level,
   # then reposition to just after the last child of the window.
   if ($cur->toplevel() == $cur)
    {
     $parent = $cur;
     @children = $cur->FocusChildren();
     $i = @children;
    }
   else
    {
     $parent = $cur->parent();
     @children = $parent->FocusChildren();
     $i = lsearch(\@children,$cur);
    }
   # Go to the previous sibling, then descend to its last descendant
   # (highest in stacking order. While doing this, ignore top-levels
   # and their descendants. When we run out of descendants, go up
   # one level to the parent.
   while ($i > 0)
    {
     $i--;
     $cur = $children[$i];
     next if ($cur->toplevel() == $cur);
     $parent = $cur;
     @children = $parent->FocusChildren();
     $i = @children;
    }
   $cur = $parent;
   if ($cur == $w || $cur->FocusOK)
    {
     $cur->Tk::focus;
     return;
    }
  }

}

sub FocusOK
{
 my $w = shift;
 my $value;
 catch { $value = $w->cget('-takefocus') };
 if (!$@ && defined($value))
  {
   return 0 if ($value eq '0');
   return 1 if ($value eq '1');
   $value = $w->$value();
   return $value if (defined $value);
  }
 if (!$w->viewable)
  {
   return 0;
  }
 catch { $value = $w->cget('-state') } ;
 if (!$@ && defined($value) && $value eq "disabled")
  {
   return 0;
  }
 $value = grep(/Key|Focus/,$w->Tk::bind(),$w->Tk::bind(ref($w)));
 return $value;
}


# focusFollowsMouse
#
# If this procedure is invoked, Tk will enter "focus-follows-mouse"
# mode, where the focus is always on whatever window contains the
# mouse. If this procedure isn't invoked, then the user typically
# has to click on a window to give it the focus.
#
# Arguments:
# None.

sub EnterFocus
{
 my $w  = shift;
 my $Ev = $w->XEvent;
 my $d  = $Ev->d;
 $w->Tk::focus() if ($d eq "NotifyAncestor" ||  $d eq "NotifyNonlinear" ||  $d eq "NotifyInferior");
}

sub focusFollowsMouse
{
 my $widget = shift;
 $widget->bind('all',"EnterFocus");
}

# tkTraverseToMenu --
# This procedure implements keyboard traversal of menus. Given an
# ASCII character "char", it looks for a menubutton with that character
# underlined. If one is found, it posts the menubutton's menu
#
# Arguments:
# w - Window in which the key was typed (selects
# a toplevel window).
# char - Character that selects a menu. The case
# is ignored. If an empty string, nothing
# happens.
sub TraverseToMenu
{
 my $w = shift;
 my $char = shift;
 return unless(defined $char && $char ne "");
 $w = $w->toplevel->FindMenu($char);
 $w->PostFirst() if (defined $w);
}
# tkFirstMenu --
# This procedure traverses to the first menubutton in the toplevel
# for a given window, and posts that menubutton's menu.
#
# Arguments:
# w - Name of a window. Selects which toplevel
# to search for menubuttons.
sub FirstMenu
{
 my $w = shift;
 $w = $w->toplevel->FindMenu("");
 $w->PostFirst() if (defined $w);
}

# These wrappers don't use method syntax so need to live
# in same package as raw Tk routines are newXS'ed into.

sub Selection
{my $widget = shift;
 my $cmd    = shift;
 croak "Use SelectionOwn/SelectionOwner" if ($cmd eq 'own');
 croak "Use Selection\u$cmd()";
}

sub Clipboard
{my $w = shift;
 my $cmd    = shift;
 croak "Use clipboard\u$cmd()";
}

sub Receive
{
 my $w = shift;
 warn "Receive(" . join(',',@_) .")";
 die "Tk rejects send(" . join(',',@_) .")\n";
}

sub break
{
 die "_TK_BREAK_\n";
}

sub idletasks
{
 shift->update('idletasks');
}

sub updateWidgets
{
 my ($w) = @_;
 while ($w->DoOneEvent(DONT_WAIT|IDLE_EVENTS|WINDOW_EVENTS))
  {
  }
 $w;
}

sub ImageNames
{
 image('names');
}

sub ImageTypes
{
 image('types');
}

sub interps
{
 my $w = shift;
 return $w->winfo('interps','-displayof');
}

sub findINC
{
 my $file = join('/',@_);
 my $dir;
 $file  =~ s,::,/,g;
 foreach $dir (@INC)
  {
   my $path;
   return $path if (-e ($path = "$dir/$file"));
  }
 return undef;
}

sub lsearch
{my $ar = shift;
 my $x  = shift;
 my $i;
 for ($i = 0; $i < scalar @$ar; $i++)
  {
   return $i if ($$ar[$i] eq $x);
  }
 return -1;
}

# a wrapper on eval which turns off user $SIG{__DIE__}
sub catch (&)
{
 my $sub = shift;
 eval {local $SIG{'__DIE__'}; &$sub };
}

